--- ../q1pro-klipper-backup/klipper_config/printer.cfg	2025-02-11 07:05:00.000000000 +0800
+++ ../klipper_config/printer.cfg	2025-02-12 04:34:15.355774031 +0800
@@ -195,14 +195,15 @@
 full_steps_per_rotation: 200
 endstop_pin:probe:z_virtual_endstop #U_1:PC3 for Z-max
 endstop_pin_reverse:tmc2209_stepper_z:virtual_endstop
-position_endstop:-0.2
+#position_endstop:-0.2
 position_endstop_reverse:248
 position_max:248
 position_min: -6
 homing_speed: 8
 homing_speed_reverse: 8
 second_homing_speed: 10
-homing_retract_dist: 5.0
+#homing_retract_dist: 5.0
+homing_retract_dist: 0.0
 homing_positive_dir:false
 homing_positive_dir_reverse:true
 step_pulse_duration:0.0000001
@@ -217,19 +218,34 @@
 step_pulse_duration:0.0000001
 endstop_pin_reverse:tmc2209_stepper_z1:virtual_endstop
 
+# [z_tilt]
+# z_positions:
+#     -59,125
+#     307.5,125
+
+# points:
+#     0,125
+#     215,125
+
+# speed: 150
+# horizontal_move_z: 5
+# retries: 2
+# retry_tolerance: 0.05
+
+#these values should be OK to test with, fine tune later for forward X offset, Y should be same
 [z_tilt]
 z_positions:
     -59,125
     307.5,125
 
 points:
-    0,125
-    215,125
+    0, 125
+    215, 125
 
 speed: 150
 horizontal_move_z: 5
-retries: 2
-retry_tolerance: 0.05
+retries: 5
+retry_tolerance: 0.008
 
 [tmc2240 stepper_y]
 cs_pin:U_1:PB9
@@ -401,36 +417,58 @@
 shutdown_value:0
 value:0
 
-[smart_effector]
-pin:U_1:PC1
-recovery_time:0
-x_offset: 17.6
-y_offset: 4.4
-#z_offset: 0.000001
-speed:5
-lift_speed:5
-probe_accel:50
-samples: 2
-samples_result: submaxmin
-sample_retract_dist: 5.0
-samples_tolerance: 0.05
-samples_tolerance_retries:5
-
-[qdprobe]
-pin:!gpio21
-z_offset:0.000001
+# [smart_effector]
+# pin:U_1:PC1
+# recovery_time:0
+# x_offset: 17.6
+# y_offset: 4.4
+# #z_offset: 0.000001
+# speed:5
+# lift_speed:5
+# probe_accel:50
+# samples: 2
+# samples_result: submaxmin
+# sample_retract_dist: 5.0
+# samples_tolerance: 0.05
+# samples_tolerance_retries:5
+
+# [qdprobe]
+# pin:!gpio21
+# z_offset:0.000001
+
+[beacon]
+serial: /dev/serial/by-id/usb-Beacon_Beacon_RevH_5374DE4C5157355957202020FF123126-if00
+x_offset: 9                     # update with X offset from nozzle on your machine
+y_offset: 0                # update with Y offset from nozzle on your machine
+mesh_main_direction: x
+mesh_runs: 2
+contact_max_hotend_temperature: 180
+home_xy_position: 120, 120      # update with your safe Z home position
+home_z_hop: 5
+home_z_hop_speed: 30
+home_xy_move_speed: 300
+home_y_before_x: False
+home_method: contact
+home_method_when_homed: proximity
+home_autocalibrate: unhomed
+home_gcode_pre_x: _BEACON_HOME_PRE_X
+home_gcode_post_x: _BEACON_HOME_POST_X
+home_gcode_pre_y: _BEACON_HOME_PRE_Y
+home_gcode_post_y: _BEACON_HOME_POST_Y
+contact_activate_gcode: _BEACON_CONTACT_PRE_Z
+contact_deactivate_gcode: _BEACON_CONTACT_POST_Z
 
 [bed_mesh]
-speed:150 
-horizontal_move_z:7
+speed:250 
+horizontal_move_z:2
 mesh_min:20,15      
-mesh_max:230,230    
-probe_count:8,8     
+mesh_max:215,230    
+probe_count:15,15     
 algorithm:bicubic
 bicubic_tension:0.2
 mesh_pps: 2, 2
-vibrate_gcode:
-    Z_DOUDONG
+#vibrate_gcode:
+#    Z_DOUDONG
 
 [filament_switch_sensor fila]
 pause_on_runout: True
@@ -441,7 +479,7 @@
 switch_pin:U_1:PC3
 
 [resonance_tester]
-accel_chip:adxl345
+accel_chip:beacon #adxl345
 probe_points:
    120, 120, 10
 

